### PR TITLE
Introducing an orderbook class (typescript)

### PR DESCRIPTION
**Pull request by fleupold**
The goal of this PR is to lay the foundation of allowing typesafe logic to be added to the dex-contracts repo. The first use case will be an Orderbook class, which can be created from bids and asks and for now simply serialized into a json format that can be easily visualized (it's the same format I'm using here: https://6jxh4.csb.app/)

Going forward this class will receive logic such as inverting an orderbook (bids become asks and vice versa), "adding" orderbooks (combining two books on the same pair into one) and "multiplying" (turning A:B & B:C into A:C). All this will be useful for visualization of order data as well as for estimating what limit price an order should be using in order to have a good chance of being matched.

The typescript compilation and tests are run as part of travis and we will be able to depend on the compiled js resources in our js scripts.

### Test Plan

`yarn run test-ts` and making sure that after running yarn build we can depend on `/typescript/orderbook.js` in the good old javascript truffle scripts.

### CODE REVIEW

File: src/orderbook.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+export class Orderbook {
+  from: string;
```
Comment (anxolin) [inline src/orderbook.ts]: nit: What is this from? By the example data in the test, it looks like it was used for the token symbol.
If it's just a label, maybe something like `sellLabel` or `sellTokenLabel`.
If it's an address, `sellTokenAddress`

File: src/orderbook.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  to: string;
+  asks: Map<number, number>;
```
Comment (anxolin) [inline src/orderbook.ts]: I think it's better to make these two properties private

File: src/orderbook.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  addAsk(price: number, volume: number) {
+    addOffer(price, volume, this.asks);
+  }
+
+  toJSON(): string {
+    const asks = Array.from(this.asks.entries());
+    const bids = Array.from(this.bids.entries());
+    asks.sort((lhs, rhs) => lhs[0] - rhs[0]);
+    bids.sort((lhs, rhs) => lhs[0] - rhs[0]);
+    return JSON.stringify({bids, asks});
```
Comment (anxolin) [inline src/orderbook.ts]: It looks the `toJSON` does 2 things:
- get the book in a representable way
- stringify it

So I would split it in two functions. Actually, I would only implement the first one, since `JSON.stringify` can be done by the user of this order book.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,67 @@
+{
```
Comment (anxolin) [inline tsconfig.json:1]: For this part, I'll refer you to our typescript master, @Velenir 

From my side, I just checked you have `strict` mode on.

Probably you should also enable ES6 output so we can later benefit from tree shaking.

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */
+
+    /* Experimental Options */
+    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
+    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
+
+    /* Advanced Options */
+    "forceConsistentCasingInFileNames": true /* Disallow inconsistently-cased references to the same file. */
+  },
+  "files": ["src/orderbook.ts"]
```
Comment (anxolin) [inline tsconfig.json]: I would use a wildcard expression OR define an entry point `src/index.ts`

**Review by anxolin — COMMENTED** [review]
Good job!

I have some thoughts about this PR:
- You added Typescript to dex-js. Good call, love that. I would love to see in the future evolving more parts of this repo to it.
- I like the new data structure. It's simple and straightforward.
- I'm not sure it's a good thing adding this in this project. I don't think this is part of the smart contract project responsibilities. I think dex-js is a better fit for it.

Added some comments.

File: src/orderbook.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  }
+}
+
+function addOffer(price: number, volume: number, map: Map<number, number>) {
```
Comment (anxolin) [inline src/orderbook.ts]: Instead of `map` can you call it sth more meaningful. Like `pricePoints`

File: test/models/orderbook.spec.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    orderbook.addBid(0.9, 50);
+    orderbook.addBid(0.99, 70);
+    orderbook.addBid(0.9, 30);
+
+    assert.equal(
+      orderbook.toJSON(),
+      JSON.stringify({
+        bids: [
+          [0.9, 80],
```
Comment (anxolin) [inline test/models/orderbook.spec.ts]: Are you sure you want to use this structure for the bid. It seems compact, but an object like approach is more readable and extendable.

```js
asks: [
          { price: 1.01, volume: 300},
          { price: 1.1, volume: 100},
          ...
]
```

Comment (anxolin) [reply test/models/orderbook.spec.ts]: One example for why would you want to use an object style representation is, if you use this for creating a endpoint or a graph, you might want to also show how many orders are in a price point, or any other relevant information.

File: src/orderbook.ts (hunk 3 above)
Comment (anxolin) [reply src/orderbook.ts]: Also, I'm not a big fan of acronyms like lhs and rhs. 

what about:
```js
Array.from(this.bids.entries())
   .sort(([leftPrice], [rightPrice]) => leftPrice - rightPrice)
```

Also, maybe u can create a function so you can reuse it for both sortings.

File: src/orderbook.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  bids: Map<number, number>;
+
+  constructor(from: string, to: string) {
+    this.from = from;
+    this.to = to;
+    this.asks = new Map();
+    this.bids = new Map();
+  }
+
+  addBid(price: number, volume: number) {
```
Comment (josojo) [inline src/orderbook.ts]: Should the price be numerator and denominator? Can we just represent the price with one number?

Probably, we would save us a lot of hassle, if we would not use floats but exact numbers.

**Review by josojo — APPROVED** [review]
LGTM

File: tsconfig.json (hunk 4 above)
Comment (Velenir) [reply tsconfig.json:1]: Yes, would be nice to have two bundles: default with `module: commonjs` , and treeshakable with `module: es6`
Take [build script](https://github.com/gnosis/dex-js/blob/master/package.json#L14) as a reference

Another thing to add for treeshaking is [all these fields](https://github.com/gnosis/dex-js/blob/master/package.json#L34-L38)

File: src/orderbook.ts (hunk 1 above)
Comment (fleupold) [reply src/orderbook.ts]: Agree that this naming is not very clear. However in an orderbook the notion of a sell/buy token does not really exist as the same token can be sold (to bidders) and bought (from askers).

I think the current name is base currency and quote currency (cf https://support.kraken.com/hc/en-us/articles/115000364388-Trading-glossary). Will change

File: src/orderbook.ts (hunk 3 above)
Comment (fleupold) [reply src/orderbook.ts]: 👍

File: tsconfig.json (hunk 5 above)
Comment (fleupold) [reply tsconfig.json]: `src/*.ts` didn't work. I'd suggest we add the entrypoint when we actually have more than one ts file

File: test/models/orderbook.spec.ts (hunk 7 above)
Comment (fleupold) [reply test/models/orderbook.spec.ts]: The reason I used this representation is that it is the one used in the visualization library (https://6jxh4.csb.app/). It's also what the [poliniex api](https://poloniex.com/public?command=returnOrderBook&currencyPair=BTC_ETH&depth=50) returns. I agree that naming price/volume can be useful for comprehension and type safety. 

I don't think the number of orders at a given price point is relevant for an order book (the depth is).

File: src/orderbook.ts (hunk 8 above)
Comment (fleupold) [reply src/orderbook.ts]: At some point we should map the price to a single number as it makes practical calculation much simpler. We can leave it as a fraction internal this class in order to avoid rounding errors e.g. when inverting order books.

File: package.json [hunk 9]
```diff
@@ -61,13 +62,19 @@
   },
   "homepage": "https://github.com/gnosis/dex-contracts#readme",
   "dependencies": {
-    "bn.js": "^5.1.1"
+    "@types/chai": "^4.2.10",
+    "@types/mocha": "^7.0.2",
+    "bn.js": "^5.1.1",
+    "chai": "^4.2.0",
+    "mocha": "^7.1.0",
+    "ts-node": "^8.6.2"
```
Comment (nlordell) [inline package.json]: I think all these new dependencies should be `devDependencies`

File: src/orderbook.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+export class Orderbook {
+  from: string;
+  to: string;
+  asks: Map<number, number>;
+  bids: Map<number, number>;
```
Comment (nlordell) [inline src/orderbook.ts]: Any reason why you are using the map type instead of a plain old JS object:
```
asks: { [key: number]: number }
bids: { [key: number]: number }
```

Just curious, I'm 100% fine with how it is.

File: tsconfig.json (hunk 5 above)
Comment (nlordell) [reply tsconfig.json]:
```suggestion
  "include": ["src/**/*.ts"],
```
Using `include` instead of `files` allows you to use wildcards

File: src/orderbook.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    this.to = to;
+    this.asks = new Map();
+    this.bids = new Map();
+  }
+
+  addBid(price: number, volume: number) {
+    addOffer(price, volume, this.bids);
+  }
+
+  addAsk(price: number, volume: number) {
```
Comment (nlordell) [inline src/orderbook.ts]: ~~Should volume use `BN` because of precision issues?~~

*edit*: I guess it already takes the decimals into account - my bad.

File: tsconfig.json (hunk 5 above)
Comment (anxolin) [reply tsconfig.json]: Okok

File: src/orderbook.ts (hunk 1 above)
Comment (anxolin) [reply src/orderbook.ts]: okay, look good to me if you change it to those names you suggest. 
Just not "from/to" cause it can be very confusing, specially in crypto.

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es5" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
+    "module": "commonjs" /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */,
+    // "lib": [],                             /* Specify library files to be included in the compilation. */
+    // "allowJs": true,                       /* Allow javascript files to be compiled. */
+    // "checkJs": true,                       /* Report errors in .js files. */
+    // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
+    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
+    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
+    // "sourceMap": true,                     /* Generates corresponding '.map' file. */
```
Comment (nlordell) [inline tsconfig.json:13]: Maybe enable source maps ~~(I suggest with the `inlineSourceMap` option)~~ so that assertion failures for tests show Typescript line numbers and are easier to follow.

**edit**: plain old source maps are fine.

File: test/models/orderbook.spec.ts (hunk 7 above)
Comment (anxolin) [reply test/models/orderbook.spec.ts]: Not a big deal for me. If it's simpler for your targeted use case, I'm good

Comment (anxolin) [reply test/models/orderbook.spec.ts]: @fleupold this is for taking a snapshot of the orderbook at our current moment in time, right?

We don't want to take validity into account. I mean, a more advanced version of this could keep track of the orders and return the book for a given batch. I understand that's not your current goal.

It can also be a service  that returns the book for a given batchId (future or past)

**Review by anxolin — APPROVED** [review]

File: tsconfig.json (hunk 12 above)
Comment (Velenir) [reply tsconfig.json:13]: Please don't inline source maps, frontend will be sad.
If no .js files will end up in the frontend (currently we only import `BatchExchange.json` through dex-js ), then `inlineSourceMap` is ok

Comment (nlordell) [reply tsconfig.json:13]: I don't think any of these JS files would get added to the front end with how things are currently imported. `index.js` just imports the contracts ABIs and the encoding methods as of right now.

Also, just curious, @Velenir what are the issues with inlined source maps with the front end? Don't they get stripped on minification? Also, shouldn't JS bundling tools also work with transitive source maps (as in the bundler and minifier would produce a source map, which would map the minified bundle to the original source by reading the intermediate inlined source map)?

Comment (Velenir) [reply tsconfig.json:13]: No issue with minification, bundling. It works exactly as you described

There are some seemingly random issues when Chrome and/or Firefox devtools sometimes can't read sourcemaps properly and debugging 3rd-party libraries gets difficult.

File: src/orderbook.ts (hunk 3 above)
Comment (fleupold) [reply src/orderbook.ts]: Turns out toJSON should not serialize, just return the struct you want later to be serialized (https://github.com/douglascrockford/JSON-js/blob/master/json2.js). Will change it so that we return the struct and leave it up to the caller to serialize.

File: test/models/orderbook.spec.ts (hunk 7 above)
Comment (fleupold) [reply test/models/orderbook.spec.ts]: It depends what is your input here. If you are e.g. using the BatchExchangeViewer.sol contract you can chose for which "target batch" you want your orders and availableSellBalances. This way you can make it behave like you suggested. 
The "orderbook logic" is independent of its data source (could even come from the graph)

File: src/orderbook.ts (hunk 10 above)
Comment (fleupold) [reply src/orderbook.ts]: Map has a bunch of convenience functions such as entries, values, get and is a little bit more concise to write that the good old dictionary type def.

File: .travis.yml [hunk 13]
```diff
@@ -16,7 +16,8 @@ script:
   - yarn run pretty-check
   - solium -d contracts/
   - yarn run coverage && cat ./coverage/lcov.info | coveralls
-  - yarn run test ./test/stablex/stablex_large_example.js
+  - npx truffle test ./test/stablex/stablex_large_example.js
```
Comment (nlordell) [inline .travis.yml]: ubernit: Is it possible to make this an `package.json` `script` as well?

File: package.json [hunk 14]
```diff
@@ -24,7 +27,9 @@
     "prettier:js": "prettier --write './**/*.js'",
     "pretty-check": "prettier --check 'contracts/**/*.sol' && prettier --check './**/*.js'",
     "spread-orders": "npx truffle exec scripts/stablex/place_spread_orders.js",
-    "ts": "yarn run ts-contracts && yarn run ts-check && tsc",
+    "ts": "yarn run ts-contracts && yarn run ts-check && yarn run tsc:commonjs && yarn run tsc:esm",
+    "tsc:commonjs": "rimraf build && tsc",
+    "tsc:esm": "rimraf typescript-esm && tsc -m es6 -t esnext --outDir typescript-esm",
```
Comment (nlordell) [inline package.json]: ubernit: is it possible to have the output be in `typescript/common` and `typescript/esm` so that there is only one directly in the `.gitignore`. IMO its a little cleaner that way.

File: yarn.lock [hunk 15]
```diff
@@ -8413,7 +8413,11 @@ source-map-support@^0.4.15:
   dependencies:
     source-map "^0.5.6"
 
+<<<<<<< HEAD
 source-map-support@^0.5.6:
+=======
+source-map-support@^0.5.16:
+>>>>>>> origin/master
```
Comment (nlordell) [inline yarn.lock]: Merge error.

**Review by nlordell — APPROVED** [review]

File: package.json [hunk 16]
```diff
@@ -3,15 +3,20 @@
   "version": "0.2.4",
   "description": "Contracts for dFusion multi-token batch auction exchange",
   "main": "src/index.js",
+  "module": "typescript-esm/index.js",
+  "jsnext:main": "typescript-esm/index.js",
+  "types": "typescript-esm/index.d.ts",
```
Comment (nlordell) [inline package.json]: I think these paths need to be updated.

File: package.json [hunk 17]
```diff
@@ -23,7 +28,9 @@
     "prettier:js": "prettier --write './**/*.js'",
     "pretty-check": "prettier --check 'contracts/**/*.sol' && prettier --check './**/*.js'",
     "spread-orders": "npx truffle exec scripts/stablex/place_spread_orders.js",
-    "ts": "yarn run ts-contracts && yarn run ts-check",
+    "ts": "yarn run ts-contracts && yarn run ts-check && yarn run tsc:commonjs && yarn run tsc:esm",
+    "tsc:commonjs": "rimraf build && tsc --outDir typescript/common",
+    "tsc:esm": "rimraf typescript-esm && tsc -m es6 -t esnext --outDir typescript/esm",
```
Comment (nlordell) [inline package.json]: here as sell.

File: src/orderbook.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    const asks = Array.from(this.asks.values());
+    const bids = Array.from(this.bids.values());
+    asks.sort(sortOffersAscending);
+    bids.sort(sortOffersAscending);
+    bids.reverse();
+    return {bids, asks};
+  }
+}
+
+function addOffer(offer: Offer, existingOffers: Map<number, Offer>) {
```
Comment (anxolin) [inline src/orderbook.ts:64]: Should this be private?

File: src/orderbook.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+  addBid(bid: Offer) {
+    addOffer(bid, this.bids);
+  }
+
+  addAsk(ask: Offer) {
+    addOffer(ask, this.asks);
+  }
+
+  toJSON() {
```
Comment (anxolin) [inline src/orderbook.ts:54]: nit: Why the name is `toJSON`. I'd suggest to give a functional name and type to the object containing bids and asks, for example BookOffers and call the function `getBookOffers`

File: src/orderbook.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
+export class Price {
+  numerator: number;
```
Comment (anxolin) [inline src/orderbook.ts:2]: Shouldn't this bi a `BN`?

File: scripts/stablex/utilities.js [hunk 21]
```diff
@@ -143,6 +143,36 @@ const sendLiquidityOrders = async function(
...
+// Given an orderbook A:B & B:C compute orderbook B:C
+const transitiveOrderbook = function(lhs, rhs) {
+  const result = { bids: [], asks: [] }
+  const lhs_asks = lhs.asks.values()
+  const rhs_asks = rhs.asks.values()
+
+  let rhs_next = rhs_asks.next()
+  let lhs_next = lhs_asks.next()
+  while (!(lhs_next.done || rhs_next.done)) {
+    const price = lhs_next.value[0] * rhs_next.value[0]
```
Comment (anxolin) [inline scripts/stablex/utilities.js:155]: nit: I personally find it easier to read if you create a constant with a semantic name for `lhs_next.value[0]` or `lhs_next.value[1]` and include some extra comments in plain english for example for the if predicate or their branches.

File: scripts/stablex/utilities.js [hunk 22]
```diff
@@ -143,6 +143,36 @@ const sendLiquidityOrders = async function(
   )
 }
 
+// Given an orderbook A:B & B:C compute orderbook B:C
+const transitiveOrderbook = function(lhs, rhs) {
```
Comment (anxolin) [inline scripts/stablex/utilities.js:147]: I think this should be done in typescript

**Review by anxolin — COMMENTED** [review]
@fleupold I'm taking a look to the old PRs for the order book to review the price calculation, sorry for doing a second review, but I saw there was new commits since I reviewed last time.

File: src/orderbook.ts (hunk 19 above)
Comment (fleupold) [reply src/orderbook.ts:54]: toJSON is the function that JSON.stringify calls on the objects it's trying to serialize. Thus we need to implement it in order to ensure the serialization happens in the way we want to. I can create a getBookOffers method on top of that, but it wasn't needed until now.

File: src/orderbook.ts (hunk 20 above)
Comment (fleupold) [reply src/orderbook.ts:2]: By now this has become a fraction class where numerator/denominator are both BN

File: scripts/stablex/utilities.js (hunk 22 above)
Comment (fleupold) [reply scripts/stablex/utilities.js:147]: Yes, this slipped in by accident. It was my first draft at the algorithm until I decided to do all of it in ts. Will send a cleanup PR to remove (it's unused)

File: src/orderbook.ts (hunk 19 above)
Comment (anxolin) [reply src/orderbook.ts:54]: fair enough

File: src/orderbook.ts (hunk 20 above)
Comment (anxolin) [reply src/orderbook.ts:2]: Yes, I saw it later! I like more the new approach!

File: package.json [hunk 23]
```diff
@@ -3,15 +3,20 @@
...
   },
   "scripts": {
     "build": "truffle build && yarn run networks-reset && yarn run ts",
-    "test": "truffle test",
-    "test-stablex": "truffle test test/stablex/*",
-    "test-snapp": "truffle test test/snapp/*",
-    "coverage": "truffle run coverage --network coverage",
+    "test": "truffle test && yarn run test-ts",
+    "test-stablex": "yarn run test-js test/stablex/*",
+    "test-snapp": "yarn run test-js test/snapp/*",
```
Comment (bh2smith) [inline package.json:16]: I thought the snapp contracts would have been removed by now!

File: test/models/orderbook.spec.ts [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    orderbook.addBid(new Offer(new Price(9, 10), 50));
+    orderbook.addBid(new Offer(new Price(99, 100), 70));
+    orderbook.addBid(new Offer(new Price(9, 10), 30));
+
+    assert.equal(
+      JSON.stringify(orderbook.toJSON()),
+      JSON.stringify({
+        bids: [
+          {price: 0.99, volume: 70},
+          {price: 0.9, volume: 80}
```
Comment (bh2smith) [inline test/models/orderbook.spec.ts:21]: I see that identical bids are aggregated by accumulating the volume. I suppose this was the reason for using the fraction class (so that prices could be in a reduced format). In this case we might want to consider including a unit test with `Price(F_n, F_{n-1})` where the F_i's are Fibonacci numbers and `n` is large enough so that `F_n` is as close as possible to the largest number in our system (i.e. maxU256), because this will yield the longest possible run time for computing the GCD.

**Review by bh2smith — COMMENTED** [review]
Very cool stuff. One comment that is not urgent, but I thought should be noted somewhere.
